### PR TITLE
Allow `delta_t` to be array in SPA functions

### LINKED ISSUE #2189: `solarposition.spa_python` does not accept arrays for `delta_t` parameter
`delta_t`, the difference between Terrestrial Time and UT1, varies over the decades.  The default value in `solarposition.spa_python` (the scalar `delta_t=67.0`) is not suitable for high-precision calculations and/or simulations of times far from ~2010.  

Users concerned with such cases are likely to want to input their own time-varying `delta_t`.  Unfortunately that is not currently possible; `spa_python` accepts only scalar `delta_t` inputs.  The only alternative is to set `delta_t=None`, which triggers an internal time-varying calculation of `delta_t`, but this assumes a particular model for `delta_t` which may not satisfy the need.

I suggest we modify `spa_python` so that `delta_t` is allowed to be an array of the same length as the other array inputs.  The offending line of code is here:

https://github.com/pvlib/pvlib-python/blob/a2abe3e2d12d5a17411ca408f545b0bdfba1db15/pvlib/solarposition.py#L381

When `delta_t` is an array, this line raises `ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()`.  Of course this line can be easily changed to allow array `delta_t`. I'm guessing this is the only change needed to get `how='numpy'` working.  For `how='numba'`, it would be trickier (https://github.com/pvlib/pvlib-python/pull/240#issuecomment-254541563).

### PR DESCRIPTION
**Pull request by kandersolar**
- [x] Closes #2189
 - [x] I am familiar with the [contributing guidelines](https://pvlib-python.readthedocs.io/en/latest/contributing.html)
 - [x] Tests added
 - [ ] Updates entries in [`docs/sphinx/source/reference`](https://github.com/pvlib/pvlib-python/blob/main/docs/sphinx/source/reference) for API changes.
 - [x] Adds description and name entries in the appropriate "what's new" file in [`docs/sphinx/source/whatsnew`](https://github.com/pvlib/pvlib-python/tree/main/docs/sphinx/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
 - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.
 - [x] Maintainer: Appropriate GitHub Labels (including `remote-data`) and Milestone are assigned to the Pull Request and linked Issue.

See #2189.  Additional tests are needed before this is ready for detailed review.

### CODE REVIEW

File: pvlib/spa.py [hunk 1]
```diff
@@ -835,24 +835,24 @@ def equation_of_time(sun_mean_longitude, geocentric_sun_right_ascension,
     return E
 
 
-@jcompile('void(float64[:], float64[:], float64[:,:])', nopython=True,
-          nogil=True)
-def solar_position_loop(unixtime, loc_args, out):
+@jcompile('void(float64[:], float64[:], float64[:], float64[:,:])',
+          nopython=True, nogil=True)
+def solar_position_loop(unixtime, delta_t, loc_args, out):
```
Comment (kandersolar) [inline pvlib/spa.py:840]: Whereas `delta_t` used to be a scalar that could live in `loc_args`, now it is a time-varying value (same length as `unixtime`) and thus requires a separate array input.  So a new parameter is needed, and `loc_args` shrinks by one.

File: pvlib/tests/test_solarposition.py [hunk 2]
```diff
@@ -770,14 +771,14 @@ def test__datetime_to_unixtime_units(unit, tz):
...
 @requires_pandas_2_0
+@pytest.mark.parametrize('tz', [None, 'utc', 'US/Eastern'])
 @pytest.mark.parametrize('method', [
     'nrel_numpy',
     'ephemeris',
     pytest.param('pyephem', marks=requires_ephem),
     pytest.param('nrel_numba', marks=requires_numba),
     pytest.param('nrel_c', marks=requires_spa_c),
 ])
-@pytest.mark.parametrize('tz', [None, 'utc', 'US/Eastern'])
```
Comment (kandersolar) [inline pvlib/tests/test_solarposition.py:780]: This order change was so that all the numba tests run in sequence; the previous ordering interleaved numpy & numba calls, causing unnecessary reloading of `pvlib.spa`.

File: pvlib/tests/test_solarposition.py [hunk 3]
```diff
@@ -786,8 +787,12 @@ def test_get_solarposition_microsecond_index(method, tz):
     index_ns = pd.date_range(unit='ns', **kwargs)
     index_us = pd.date_range(unit='us', **kwargs)
 
-    sp_ns = solarposition.get_solarposition(index_ns, 40, -80, method=method)
-    sp_us = solarposition.get_solarposition(index_us, 40, -80, method=method)
+    with warnings.catch_warnings():
+        # don't warn on method reload
+        warnings.simplefilter("ignore")
```
Comment (kandersolar) [inline pvlib/tests/test_solarposition.py:792]: This change is not directly related; just a tweak to suppress nuisance reloading warnings

**Review by kandersolar — COMMENTED** [review]
Ready for review.  The coverage complaints are the usual numba issue; they can be ignored.

A brief explanation: the complexity here is to accommodate numba.  With the numpy calculations, duck typing means both scalar and array `delta_t` are fine.  But with numba, we have to pick a single dtype.  This PR makes `delta_t` always an array (forcing scalars to arrays if needed) so that the numba code can deal with arrays only.  

Technically this change will increase memory use, but my informal benchmarking found no noticeable slowdown.  Considering what a computational beast the SPA is, I'm not surprised that one extra array doesn't make any measurable difference.

File: pvlib/spa.py [hunk 4]
```diff
@@ -940,13 +943,17 @@ def solar_position_numba(unixtime, lat, lon, elev, pressure, temp, delta_t,
...
         return result
 
     # split the input and output arrays into numthreads chunks
     split0 = np.array_split(unixtime, numthreads)
+    split1 = np.array_split(delta_t, numthreads)
     split2 = np.array_split(result, numthreads, axis=1)
-    chunks = [[a0, loc_args, split2[i]] for i, a0 in enumerate(split0)]
+    chunks = [
+        [a0, a1, loc_args, a2]
+        for i, (a0, a1, a2) in enumerate(zip(split0, split1, split2))
```
Comment (AdamRJensen) [inline pvlib/spa.py:955]:
```suggestion
        for a0, a1, a2 in zip(split0, split1, split2)
```

**Review by AdamRJensen — APPROVED** [review]

**Comment by AdamRJensen** [comment]
A better test instead of [65, 65] would be to have an array input with different values, e.g., [65, 66].
